src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -1690,7 +1693,7 @@ private ScriptBlockAst NamedBlockListRule(Token lCurly, List<UsingStatementAst>
 
             IScriptExtent scriptBlockExtent;
             extent = ExtentOf(startExtent, endExtent);
-            CompleteScriptBlockBody(lCurly, ref extent, out scriptBlockExtent);
+            CompleteScriptBlockBody(lCurly, ref extent, out scriptBlockExtent, calledFromNamedBlockRule: true);
```

Comment:
Instead of reporting here, I think the error should be reported in this function before the `goto finished_named_block_list`.

And the proposed error text might be confusing in some cases, it might be better to just report `unexpected token, expected 'begin', 'process', 'end', or 'dynamicparam'.

Reply:
If we handle the error before `goto finished_named_block_list`, then we will have to duplicate some of the code from `CompleteScriptBlockBody`, for example, the code to handle `lCurly == null`, and the code to handle `blockNameToken == <eof>`.

I will make some changes to `CompleteScriptBlockBody` to make it less confusing.

Reply:
I agree you don't want to duplicate unnecessarily. What about adding a case for `TokenKind.LCurly`, then consume any other token.

In the common case, you would not see a second error because there will just be a single token there, e.g. if you are in the middle of typing `process` or whatever.

Reply:
@lzybkr I moved the error handling to `NamedBlockListRule`, can you please take another look?